## Instructions {-}

1. You may talk to a friend, discuss the questions and potential directions for solving them. However, you need to write your own solutions and code separately, and not as a group activity. 

2. Do not write your name on the assignment.

3. Write your code in the *Code* cells and your answer in the *Markdown* cells of the Jupyter notebook. Ensure that the solution is written neatly enough to understand and grade.

4. Use [Quarto](https://quarto.org/docs/output-formats/html-basics.html) to print the *.ipynb* file as HTML. You will need to open the command prompt, navigate to the directory containing the file, and use the command: `quarto render filename.ipynb --to html`. Submit the HTML file.

5. The assignment is worth 100 points, and is due on **Thursday, 26th January 2023 at 11:59 pm**. 

6. There is a **bonus** question worth 5 points.

7.  The maximum possible score in the assigment is 100 + 5 (bonus question) = 105 out of 100. There is no partial credit for bonus question.

8. **Five points are properly formatting the assignment**. The breakdown is as follows:
- Must be an HTML file rendered using Quarto (1 pt). *If you have a Quarto issue, you must mention the issue & quote the error you get when rendering using Quarto in the comments section of Canvas, and submit the ipynb file.* 
- No name can be written on the assignment, nor can there be any indicator of the student’s identity—e.g. printouts of the working directory should not be included in the final submission  (1 pt)
- There aren’t excessively long outputs of extraneous information (e.g. no printouts of entire data frames without good reason, there aren’t long printouts of which iteration a loop is on, there aren’t long sections of commented-out code, etc.) (1 pt)
- Final answers of each question are written in Markdown cells (1 pt).
- There is no piece of unnecessary / redundant code, and no unnecessary / redundant text (1 pt)

## Multiple linear regression

A study was conducted on 97 men with prostate cancer who were due to receive a radical prostatectomy. The dataset *prostate.csv* contains data on 9 measurements made on these 97 men. The description of variables can be found here: https://rafalab.github.io/pages/649/prostate.html

### 
Fit a linear regression model with `lpsa` as the response and the other variables as the predictors. Write down the equation to predict `lpsa` based on the other eight variables.

*(2+2 points)*

In [17]:
import pandas as pd
import seaborn as sns
import statsmodels.formula.api as smf
import numpy as np

In [18]:
prostate = pd.read_csv('./Datasets/prostate.csv')
predictors=list(prostate.columns)
predictors.remove('lpsa')
predictors = "+".join(predictors)
model = smf.ols(formula='lpsa~'+predictors, data=prostate).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   lpsa   R-squared:                       0.655
Model:                            OLS   Adj. R-squared:                  0.623
Method:                 Least Squares   F-statistic:                     20.86
Date:                Tue, 17 Jan 2023   Prob (F-statistic):           2.24e-17
Time:                        09:58:35   Log-Likelihood:                -99.476
No. Observations:                  97   AIC:                             217.0
Df Residuals:                      88   BIC:                             240.1
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.6693      1.296      0.516      0.607      -1.907       3.246
lcavol         0.5870      0.088      6.677      0.000       0.412       0.762
lweight        0.4545      0.170      2.673      0.009       0.117       0.792
age           -0.0196      0.011     -1.758      0.082      -0.042       0.003
lbph           0.1071      0.058      1.832      0.070      -0.009       0.223
svi            0.7662      0.244      3.136      0.002       0.281       1.252
lcp           -0.1055      0.091     -1.159      0.250      -0.286       0.075
gleason        0.0451      0.157      0.287      0.775      -0.268       0.358
pgg45          0.0045      0.004      1.024      0.309      -0.004       0.013
==============================================================================
Omnibus:                        0.235   Durbin-Watson:                   1.507
Prob(Omnibus):                  0.889   Jarque-Bera (JB):                0.026
Skew:                          -0.017   Prob(JB):                        0.987
Kurtosis:                       3.073   Cond. No.                     1.28e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.28e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The prediction equation is
lpsa = 0.669+0.587 * lcavol + 0.454l * weight − 0.020 * age + 0.107l * bph + 0.766 * svi − 0.105 * lcp + 0.045 * gleason + 0.005 * pgg45

### 
Is the regression significant at 5% level?

For the overall significance of the regression, we get $F = 20.86$ with 8 and 88 degrees of freedom. The $p$−value = 2.2 × 10−17 is extremely small indicating that the regression is highly significant.

### 
Report the $p$-values for `gleason` and `age`. What do you conclude about the significance of these variables?

The $p$-value of the `age` variable is 0.08229. This variable is not significant at 5% level, but is significant at
10% level. So, overall, it is a marginally significant variable.

The $p$-value of the `gleason` variable is 0.775. As this value is too high, this variable is insignificant.

### 
What is the 95% confidence interval for the coefficient of `age`? Can you conclude anything about its significance based on the confidence interval?

The 95% confidence interval for the coefficient of age is \[ -0.042	0.003 \]. We can see that the 95% confidence interval includes 0 and therefore, this variable is not significant at 5% level.

### 
Fit a simple linear regression on `lpsa` against `gleason`. What is the $p$-value for `gleason`?

In [13]:
model = smf.ols(formula='lpsa~gleason', data=prostate).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   lpsa   R-squared:                       0.136
Model:                            OLS   Adj. R-squared:                  0.127
Method:                 Least Squares   F-statistic:                     14.97
Date:                Tue, 17 Jan 2023   Prob (F-statistic):           0.000200
Time:                        03:03:57   Log-Likelihood:                -143.96
No. Observations:                  97   AIC:                             291.9
Df Residuals:                      95   BIC:                             297.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.5044      1.035     -1.453      0.149      -3.559       0.550
gleason        0.5898      0.152      3.869      0.000       0.287       0.892
==============================================================================
Omnibus:                        2.099   Durbin-Watson:                   0.245
Prob(Omnibus):                  0.350   Jarque-Bera (JB):                1.526
Skew:                           0.178   Prob(JB):                        0.466
Kurtosis:                       3.501   Cond. No.                         65.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 
Is the predictor `gleason` statistically significant in the model developed in the previous question *(where it is the only predictor)? 

Was `gleason` statistically significant in the model developed in the first question with multiple predictors?

Did the statistical significance of `gleason` change in the absence of other predictors? Why or why not?

The predictor `gleason` appears to be statisticaly significant in the simple linear regression model, but insignificant in the multiple linear regression model. 

This is not surprising as the hypotheses being tested in the two models are different.

For the simple linear regression model, the hypothesis being tested is:

$H_0 : E(y) = \beta_0$ against $H_1 : E(y) = \beta_0 + \beta_1$`gleason`,

whereas in the multiple linear regression, the hypothesis being tested is:

$H_0 : E(y) = \beta_0+ \beta_1$ `lcavol`+ $\beta_2$ `lweight` + \beta_3age + \beta4lbph + \beta5svi + \beta6lcp+ \beta8pgg45$ against $H_1 : E(y) = \beta_0+ \beta_1lcavol+\beta_2lweight+\beta_3age+\beta4lbph+\beta5svi + \beta6lcp+ \beta7gleason+ \beta8pgg45$

We can further explain this observation with the help of the correlation table below.

In [14]:
prostate.corr()

lcavol   lweight       age      lbph       svi       lcp   gleason  \
lcavol   1.000000  0.194128  0.225000  0.027350  0.538845  0.675311  0.432417   
lweight  0.194128  1.000000  0.307525  0.434932  0.108778  0.100239 -0.001283   
age      0.225000  0.307525  1.000000  0.350186  0.117658  0.127668  0.268892   
lbph     0.027350  0.434932  0.350186  1.000000 -0.085843 -0.006999  0.077820   
svi      0.538845  0.108778  0.117658 -0.085843  1.000000  0.673111  0.320412   
lcp      0.675311  0.100239  0.127668 -0.006999  0.673111  1.000000  0.514830   
gleason  0.432417 -0.001283  0.268892  0.077820  0.320412  0.514830  1.000000   
pgg45    0.433652  0.050846  0.276112  0.078460  0.457648  0.631528  0.751905   
lpsa     0.734460  0.354122  0.169593  0.179810  0.566218  0.548813  0.368987   

            pgg45      lpsa  
lcavol   0.433652  0.734460  
lweight  0.050846  0.354122  
age      0.276112  0.169593  
lbph     0.078460  0.179810  
svi      0.457648  0.566218  
lcp      0.631528  0.548813  
gleason  0.751905  0.368987  
pgg45    1.000000  0.422316  
lpsa     0.422316  1.000000

Note that `gleason` is somewhat correlated with `lcavol`, and `lcavol` is highly statisticaly significant. In the simple linear regression model, the coefficient of `gleason` represents the average change in `lpsa` with a unit change in `gleason`, ignoring other predictors such as `lcavol`. However, in the multiple linear regression model, the coefficient of `gleason` represents the average change in `lpsa` with a unit change in `gleason` holding the other predictors such as `lcavol` constant. This shows that `gleason`, when by itself, is a surrogate for other statistically significant predictors such ad `lcavol`, and actually does not have a statistically significant relationship with `lpsa`.

### 
Predict `lpsa` of a 65-year old man with `lcavol` = 1.35, `lweight` = 3.65, `lbph` = 0.1, `svi` = 0.22, `lcp` = -0.18, `gleason` = 6.75, and `pgg45` = 25 and find 95% prediction intervals.

In [15]:
predictor_vals=pd.DataFrame({'lcavol':[1.35], 'lweight': 3.65, 'lbph': 0.1,'svi': [0.22],'lcp': [-0.18],'gleason': [6.75],'pgg45': [25],'age':65})
model.predict(predictor_vals)

0    2.476867
dtype: float64

In [16]:
intervals = model.get_prediction(predictor_vals)
intervals.summary_frame(alpha=0.05)

mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  2.476867  0.109506       2.259469       2.694264      0.324755   

   obs_ci_upper  
0      4.628979

Predicted value for `lpsa` = 2.46, and prediction interval is \[1.04, 3.87\]

In [34]:
prostate = pd.read_csv('./Datasets/prostate.csv')
predictors=list(prostate.columns)
predictors.remove('lpsa')
predictors = "+".join(predictors)
model = smf.ols(formula='lpsa~'+predictors, data=prostate).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   lpsa   R-squared:                       0.655
Model:                            OLS   Adj. R-squared:                  0.623
Method:                 Least Squares   F-statistic:                     20.86
Date:                Tue, 17 Jan 2023   Prob (F-statistic):           2.24e-17
Time:                        10:02:55   Log-Likelihood:                -99.476
No. Observations:                  97   AIC:                             217.0
Df Residuals:                      88   BIC:                             240.1
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.6693      1.296      0.516      0.607      -1.907       3.246
lcavol         0.5870      0.088      6.677      0.000       0.412       0.762
lweight        0.4545      0.170      2.673      0.009       0.117       0.792
age           -0.0196      0.011     -1.758      0.082      -0.042       0.003
lbph           0.1071      0.058      1.832      0.070      -0.009       0.223
svi            0.7662      0.244      3.136      0.002       0.281       1.252
lcp           -0.1055      0.091     -1.159      0.250      -0.286       0.075
gleason        0.0451      0.157      0.287      0.775      -0.268       0.358
pgg45          0.0045      0.004      1.024      0.309      -0.004       0.013
==============================================================================
Omnibus:                        0.235   Durbin-Watson:                   1.507
Prob(Omnibus):                  0.889   Jarque-Bera (JB):                0.026
Skew:                          -0.017   Prob(JB):                        0.987
Kurtosis:                       3.073   Cond. No.                     1.28e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.28e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [35]:
model.mse_resid

0.5018525374082506

In [37]:
model = smf.ols(formula='lpsa~lcavol+lweight+svi', data=prostate).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   lpsa   R-squared:                       0.626
Model:                            OLS   Adj. R-squared:                  0.614
Method:                 Least Squares   F-statistic:                     51.99
Date:                Tue, 17 Jan 2023   Prob (F-statistic):           8.05e-20
Time:                        10:03:08   Log-Likelihood:                -103.30
No. Observations:                  97   AIC:                             214.6
Df Residuals:                      93   BIC:                             224.9
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.2681      0.543     -0.493      0.623      -1.347       0.811
lcavol         0.5516      0.075      7.388      0.000       0.403       0.700
lweight        0.5085      0.150      3.386      0.001       0.210       0.807
svi            0.6662      0.210      3.176      0.002       0.250       1.083
==============================================================================
Omnibus:                        0.647   Durbin-Watson:                   1.467
Prob(Omnibus):                  0.723   Jarque-Bera (JB):                0.776
Skew:                          -0.118   Prob(JB):                        0.679
Kurtosis:                       2.631   Cond. No.                         31.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [38]:
model.mse_resid

0.513815701532462

In [31]:
hypotheses = '(gleason = 0, pgg45 = 0, lcp=0, age=0, lbph=0)'
f_test = model.f_test(hypotheses)
print(f_test)

<F test: F=array([[3.15970915]]), p=0.007461102808167509, df_denom=88, df_num=6>


## Petrol consumption

Read the dataset *petrol_consumption_train.csv*. It contains the following five columns: 

`Petrol_tax`: Petrol tax (cents per gallon) 

`Per_capita_income`: Average income (dollars) 

`Paved_highways`: Paved Highways (miles) 

`Prop_license`: Proportion of population with driver's licenses 

`Petrol_consumption`: Consumption of petrol (millions of gallons)